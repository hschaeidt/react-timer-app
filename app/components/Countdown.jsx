import React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';

export const STATUS = {
  STARTED: 'started',
  STOPPED: 'stopped',
};

export default class Countdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        case STATUS.STOPPED:
          clearInterval(this.timer);
          break;
      }
    }
  }

  startTimer() {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.setState({
          status: STATUS.STOPPED,
        });
      }
    }, 1000);
  }

  handleSetCountdown(seconds) {
    this.setState({
      count: seconds,
      status: STATUS.STARTED,
    });
  }

  render() {
    const { count } = this.state;

    return (
      <div>
        <Clock totalSeconds={count} />
        <CountdownForm onSetCountdown={this.handleSetCountdown} />
      </div>
    );
  }
}
