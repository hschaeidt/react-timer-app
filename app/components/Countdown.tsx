import * as React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';
import Controls from './Controls';

export enum STATUS {
  STARTED,
  STOPPED,
  PAUSED,
}

interface CountdownProps extends React.Props<any> {}

interface CountdownState {
  count: number,
  status: STATUS,
}

export default class Countdown extends React.Component<CountdownProps, CountdownState> {
  private timer: number;

  constructor(props: CountdownProps) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
    this.handleStatusChange = this.handleStatusChange.bind(this);
  }

  componentDidUpdate(prevProps: CountdownProps, prevState: CountdownState): void {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        default:
          this.stopTimer();
          break;
      }
    }
  }

  componentWillUnmount(): void {
    this.stopTimer();
  }

  startTimer(): void {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.handleStatusChange(STATUS.STOPPED);
      }
    }, 1000);
  }

  stopTimer(): void {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  handleSetCountdown(seconds: number) {
    if (seconds > 0) {
      this.setState({
        count: seconds,
        status: STATUS.STARTED,
      });
    }
  }

  handleStatusChange(status: STATUS) {
    const state: CountdownState = { 
      status,
      count: this.state.count,
    };

    if (status === STATUS.STOPPED) {
      state.count = 0;
    }

    this.setState(state);
  }

  render() {
    const { count, status } = this.state;

    const renderControlArea = () => {
      if (status !== STATUS.STOPPED) {
        return <Controls countdownStatus={status} onStatusChange={this.handleStatusChange} />;
      }

      return <CountdownForm onSetCountdown={this.handleSetCountdown} />;
    };

    return (
      <div>
        <h1 className="page-title">Countdown App</h1>
        <Clock totalSeconds={count} />
        {renderControlArea()}
      </div>
    );
  }
}
