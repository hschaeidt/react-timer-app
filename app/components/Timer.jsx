import React from 'react';
import Clock from './Clock';
import Controls from './Controls';
import { STATUS } from './Countdown';

export default class Timer extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.handleStatusChange = this.handleStatusChange.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.status !== prevState.status) {
      if (this.state.status === STATUS.STARTED) {
        this.startTimer();
      } else {
        this.stopTimer();
      }
    }
  }

  componentWillUnmount() {
    this.stopTimer();
  }

  startTimer() {
    this.timer = setInterval(() => {
      const count = this.state.count + 1;
      this.setState({ count });
    }, 1000);
  }

  stopTimer() {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  handleStatusChange(status) {
    const state = { status };

    if (status === STATUS.STOPPED) {
      state.count = 0;
    }

    this.setState(state);
  }

  render() {
    const { count, status } = this.state;

    return (
      <div>
        <h1 className="page-title">Timer App</h1>
        <Clock totalSeconds={count} />
        <Controls countdownStatus={status} onStatusChange={this.handleStatusChange} />
      </div>
    );
  }
}
