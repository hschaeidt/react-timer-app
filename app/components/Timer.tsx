import * as React from 'react';
import Clock from './Clock';
import Controls from './Controls';
import { STATUS } from './Countdown';

interface TimerProps extends React.Props<any> {}
interface TimerState {
  count: number;
  status: STATUS;
}

export default class Timer extends React.Component<TimerProps, TimerState>  {
  private timer: number;

  constructor(props: TimerProps) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.handleStatusChange = this.handleStatusChange.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
  }

  componentDidUpdate(prevProps: TimerProps, prevState: TimerState) {
    if (this.state.status !== prevState.status) {
      if (this.state.status === STATUS.STARTED) {
        this.startTimer();
      } else {
        this.stopTimer();
      }
    }
  }

  componentWillUnmount() {
    this.stopTimer();
  }

  startTimer(): void {
    this.timer = setInterval(() => {
      const count = this.state.count + 1;
      this.setState({ count });
    }, 1000);
  }

  stopTimer(): void {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  handleStatusChange(status: STATUS): void {
    const state: TimerState = {
      status,
      count: this.state.count,
    };

    if (status === STATUS.STOPPED) {
      state.count = 0;
    }

    this.setState(state);
  }

  render(): JSX.Element {
    const { count, status } = this.state;

    return (
      <div>
        <h1 className="page-title">Timer App</h1>
        <Clock totalSeconds={count} />
        <Controls countdownStatus={status} onStatusChange={this.handleStatusChange} />
      </div>
    );
  }
}
