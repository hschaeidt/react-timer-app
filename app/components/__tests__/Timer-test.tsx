import * as React from 'react';
import { shallow } from 'enzyme';
import * as renderer from 'react-test-renderer';
import Timer from '../Timer';
import { STATUS } from '../Countdown';

jest.useFakeTimers();

function setup() {
  const wrapper = shallow(<Timer />);
  const instance = wrapper.instance();
  const tree = renderer.create(<Timer />);

  return {
    wrapper,
    instance,
    tree,
  };
}

describe('<Timer />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Timer />);
    expect(tree).toMatchSnapshot();
  });

  it('should change the status to started', () => {
    const { instance } = setup();
    instance.handleStatusChange(STATUS.STARTED);
    expect(instance.state.status).toBe(STATUS.STARTED);
  });

  it('should stop the timer and reset the count', () => {
    const { instance } = setup();
    instance.handleStatusChange(STATUS.STARTED);
    jest.runOnlyPendingTimers();
    expect(instance.state.count).toBe(1);
    instance.handleStatusChange(STATUS.STOPPED);
    expect(instance.state.status).toBe(STATUS.STOPPED);
    expect(instance.state.count).toBe(0);
  });

  it('should clear the timer on unmount', () => {
    const { wrapper, instance } = setup();

    expect(instance.timer).not.toBeDefined();
    instance.handleStatusChange(STATUS.STARTED);
    expect(instance.timer).toBeDefined();

    wrapper.unmount();
    expect(instance.timer).not.toBeDefined();
  });
});
