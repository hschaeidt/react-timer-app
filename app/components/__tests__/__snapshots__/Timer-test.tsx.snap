exports[`<Timer /> renders correctly 1`] = `
<div>
  <h1
    className="page-title">
    Timer App
  </h1>
  <div
    className="clock">
    <span
      className="clock-text">
      00:00
    </span>
  </div>
  <div
    className="controls">
    <button
      className="button primary"
      onClick={[Function]}>
      Start
    </button>
    <button
      className="button alert hollow"
      onClick={[Function]}>
      Clear
    </button>
  </div>
</div>
`;
